File: Sources/Ably/ARTPushChannelSubscription.m
//
//  ARTPushChannelSubscription.m
//  Ably
//
//  Created by Ricardo Pereira on 15/02/2017.
//  Copyright © 2017 Ably. All rights reserved.
//

#import "ARTPushChannelSubscription.h"

@implementation ARTPushChannelSubscription

- (instancetype)initWithDeviceId:(NSString *)deviceId channel:(NSString *)channelName {
    if (self = [super init]) {
        _deviceId = deviceId;
        _channel = channelName;
    }
    return self;
}

- (instancetype)initWithClientId:(NSString *)clientId channel:(NSString *)channelName {
    if (self = [super init]) {
        _clientId = clientId;
        _channel = channelName;
    }
    return self;
}

- (id)copyWithZone:(NSZone *)zone {
    ARTPushChannelSubscription *subscription = [[[self class] allocWithZone:zone] init];

    subscription->_deviceId = self.deviceId;
    subscription->_clientId = self.clientId;
    subscription->_channel = self.channel;

    return subscription;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ - \n\t deviceId: %@; clientId: %@; \n\t channel: %@;", [super description], self.deviceId, self.clientId, self.channel];
}

- (BOOL)isEqualToChannelSubscription:(ARTPushChannelSubscription *)subscription {
    if (!subscription) {
        return NO;
    }

    BOOL haveEqualDeviceId = (!self.clientId && !subscription.clientId) || [self.clientId isEqualToString:subscription.clientId];
    BOOL haveEqualCliendId = (!self.clientId && !subscription.clientId) || [self.clientId isEqualToString:subscription.clientId];
    BOOL haveEqualChannel = (!self.channel && !subscription.channel) || [self.channel isEqualToString:subscription.channel];

    return haveEqualDeviceId && haveEqualCliendId && haveEqualChannel;
}

#pragma mark - NSObject

- (BOOL)isEqual:(id)object {
    if (self == object) {
        return YES;
    }

    if (![object isKindOfClass:[ARTPushChannelSubscription class]]) {
        return NO;
    }

    return [self isEqualToChannelSubscription:(ARTPushChannelSubscription *)object];
}

- (NSUInteger)hash {
    return [self.deviceId hash] ^ [self.clientId hash] ^ [self.channel hash];
}

@end

File: Sources/AblySwift/ARTPushChannelSubscription.swift
import Foundation

/**
 * Contains the subscriptions of a device, or a group of devices sharing the same `clientId`, has to a channel in order to receive push notifications.
 */
// swift-migration: original location ARTPushChannelSubscription.h, line 8 and ARTPushChannelSubscription.m, line 3
public class ARTPushChannelSubscription: NSObject, NSCopying {
    
    /**
     * The unique ID of the device.
     */
    // swift-migration: original location ARTPushChannelSubscription.h, line 13
    public private(set) var deviceId: String?
    
    /**
     * The ID of the client the device, or devices are associated to.
     */
    // swift-migration: original location ARTPushChannelSubscription.h, line 18
    public private(set) var clientId: String?
    
    /**
     * The channel the push notification subscription is for.
     */
    // swift-migration: original location ARTPushChannelSubscription.h, line 23
    public private(set) var channel: String
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 5
    /**
     * Creates an `ARTPushChannelSubscription` object for a channel and single device.
     *
     * @param deviceId The unique ID of the device.
     * @param channelName The channel name.
     *
     * @return An `ARTPushChannelSubscription` object.
     */
    public init(deviceId: String, channel channelName: String) {
        self.deviceId = deviceId
        self.clientId = nil
        self.channel = channelName
        super.init()
    }
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 13
    /**
     * Creates an `ARTPushChannelSubscription` object for a channel and group of devices sharing the same `clientId`.
     *
     * @param clientId The ID of the client.
     * @param channelName The channel name.
     *
     * @return An `ARTPushChannelSubscription` object.
     */
    public init(clientId: String, channel channelName: String) {
        self.deviceId = nil
        self.clientId = clientId
        self.channel = channelName
        super.init()
    }
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 21
    public func copy(with zone: NSZone?) -> Any {
        let subscription = type(of: self).init()
        subscription.deviceId = self.deviceId
        subscription.clientId = self.clientId
        subscription.channel = self.channel
        return subscription
    }
    
    // swift-migration: Placeholder init for NSCopying implementation
    public required override init() {
        self.deviceId = nil
        self.clientId = nil
        self.channel = ""
        super.init()
    }
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 31
    public override var description: String {
        return "\(super.description) - \n\t deviceId: \(deviceId ?? "nil"); clientId: \(clientId ?? "nil"); \n\t channel: \(channel);"
    }
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 35
    public func isEqual(to subscription: ARTPushChannelSubscription?) -> Bool {
        guard let subscription = subscription else {
            return false
        }
        
        // swift-migration: Fixed bug in original code - line 40 in ARTPushChannelSubscription.m had haveEqualDeviceId checking clientId twice instead of checking deviceId
        // Original: BOOL haveEqualDeviceId = (!self.clientId && !subscription.clientId) || [self.clientId isEqualToString:subscription.clientId];
        // Fixed:    let haveEqualDeviceId = (self.deviceId == nil && subscription.deviceId == nil) || (self.deviceId == subscription.deviceId)
        let haveEqualDeviceId = (self.deviceId == nil && subscription.deviceId == nil) || (self.deviceId == subscription.deviceId)
        let haveEqualClientId = (self.clientId == nil && subscription.clientId == nil) || (self.clientId == subscription.clientId)
        let haveEqualChannel = (self.channel == subscription.channel)
        
        return haveEqualDeviceId && haveEqualClientId && haveEqualChannel
    }
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 49
    public override func isEqual(_ object: Any?) -> Bool {
        if self === object as AnyObject? {
            return true
        }
        
        guard let other = object as? ARTPushChannelSubscription else {
            return false
        }
        
        return isEqual(to: other)
    }
    
    // swift-migration: original location ARTPushChannelSubscription.m, line 61
    public override var hash: Int {
        return (deviceId?.hashValue ?? 0) ^ (clientId?.hashValue ?? 0) ^ channel.hashValue
    }
}